modules/core/05-port/types/router_v2.go

```diff
@@ -5,67 +5,109 @@ package types
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-// AppRouter contains a map from module name to an ordered list of IBCModules
-// which contains all the module-defined callbacks required by ICS-26.
+const sentinelMultiPacketData = "MultiPacketData"
+
+// AppRouter contains all the module-defined callbacks required by ICS-26
 type AppRouter struct {
-	routes map[string][]IBCModule
-	sealed bool
+	routes       map[string]IBCModule
+	legacyRoutes map[string]ClassicIBCModule
+
+	// classicRoutes facilitates the consecutive calls to AddRoute for existing modules.
+	classicRoutes map[string][]ClassicIBCModule
 }
 
-// NewAppRouter creates and returns a new IBCModule application router.
 func NewAppRouter() *AppRouter {
 	return &AppRouter{
-		routes: make(map[string][]IBCModule),
+		routes:        make(map[string]IBCModule),
+		legacyRoutes:  make(map[string]ClassicIBCModule),
+		classicRoutes: make(map[string][]ClassicIBCModule),
 	}
 }
 
-// Seal prevents the Router from any subsequent route handlers to be registered.
-// Seal will panic if called more than once.
-func (rtr *AppRouter) Seal() {
-	if rtr.sealed {
-		panic(errors.New("router already sealed"))
+// AddClassicRoute adds IBCModule for a given module name. It returns the Router
+// so AddRoute calls can be linked. It will panic if the Router is sealed.
+func (rtr *AppRouter) AddClassicRoute(module string, cbs ...ClassicIBCModule) *AppRouter {
+	if !sdk.IsAlphaNumeric(module) {
+		panic(errors.New("route expressions can only contain alphanumeric characters"))
+	}
+
+	if _, ok := rtr.legacyRoutes[module]; ok {
+		panic(fmt.Errorf("route %s has already been registered", module))
 	}
-	rtr.sealed = true
-}
 
-// Sealed returns a boolean signifying if the Router is sealed or not.
-func (rtr AppRouter) Sealed() bool {
-	return rtr.sealed
+	if len(cbs) == 0 {
+		panic(errors.New("no callbacks provided"))
+	}
+
+	rtr.legacyRoutes[module] = NewLegacyIBCModule(cbs...)
+
+	return rtr
 }
 
 // AddRoute adds IBCModule for a given module name. It returns the Router
 // so AddRoute calls can be linked. It will panic if the Router is sealed.
 func (rtr *AppRouter) AddRoute(module string, cbs IBCModule) *AppRouter {
-	if rtr.sealed {
-		panic(fmt.Errorf("router sealed; cannot register %s route callbacks", module))
-	}
 	if !sdk.IsAlphaNumeric(module) {
 		panic(errors.New("route expressions can only contain alphanumeric characters"))
 	}
-	rtr.routes[module] = append(rtr.routes[module], cbs)
+
+	if _, ok := cbs.(ClassicIBCModule); ok {
+		rtr.classicRoutes[module] = append(rtr.classicRoutes[module], cbs)
+
+		// in order to facilitate having a single LegacyIBCModule, but also allowing for
+		// consecutive calls to AddRoute to support existing functionality, we can re-create
+		// the legacy module with the routes as they get added.
+		if classicRoutes, ok := rtr.classicRoutes[module]; ok && len(classicRoutes) > 1 {
+			rtr.legacyRoutes[module] = NewLegacyIBCModule(classicRoutes...)
+		}
+	} else {
```

Comment:
this part I don't understand. Is it to avoid changing the current wiring in app.go?

Reply:
essentially yes. In the app.go we do things like
```golang
appStack := ...
ibcAppRouter.AddRoute(name, appStack)
appStack = wrap(appStack)
ibcAppRouter.AddRoute(name, appStack)
```
etc.

Previously, we just built up a slice of callbacks and appended, but the idea now is that we have one type (`LegacyIBCModule`) that contains this list.

So we can either

A: construct the full list of callbacks and pass to `router.AddClassicRoute`
B: internally maintain this list and build up the `LegacyIBCModule`


This is solution B. We should be able to remove this when our app.gos use version A

Reply:
I think the app.go wiring on the router v2 added here is actually incorrect. Apps shouldn't be doing the `wrap()` step anymore

I prefer option A

Reply:
Ah I understand the issue now. If we want to test ics29 for example, currently it wraps other apps and doesn't have an api to return as a standalone app, so to use it for certain handshake functions, we still call the wrapping. I guess once everything is done, we can remove the additional add route calls

Reply:
Maybe we can open an issue to track removal of this logic

Reply:
We discussed offline ^ . We will stick with current wiring and I'll open an issue to change to the final form